### PR TITLE
x/gamm product constant invariant 

### PR DESCRIPTION
**Pull request by mconcat**
closes: #287

### CODE REVIEW

File: x/gamm/keeper/invariants.go [hunk 1]
```diff
@@ -78,3 +86,63 @@ func PoolTotalWeightInvariant(keeper Keeper, bk types.BankKeeper) sdk.Invariant
 			fmt.Sprintf("\tgamm all pool calculated and stored total weight match\n")), false
 	}
 }
+
+func genericPow(base, exp sdk.Dec) sdk.Dec {
+	if !base.GTE(sdk.OneDec().MulInt64(2)) {
```
Comment (ValarDragon) [inline x/gamm/keeper/invariants.go:91]:
```suggestion
	if !base.GTE(sdk.NewDec(2) {
```

File: x/gamm/keeper/invariants.go [hunk 2]
```diff
@@ -78,3 +86,63 @@ func PoolTotalWeightInvariant(keeper Keeper, bk types.BankKeeper) sdk.Invariant
...
+		return pow(base, exp)
+	}
+	return powApprox(sdk.OneDec().Quo(base), exp.Neg(), powPrecision)
+}
+
+func constantChange(p1, p2 types.PoolI) sdk.Dec {
```
Comment (ValarDragon) [inline x/gamm/keeper/invariants.go:97]:
```suggestion
// constantChange returns the multiplicative factor difference in the pool constant, between two different pools.
// For a Balancer pool, the pool constant is prod_{t in tokens} t.bal^{t.weight}
func constantChange(p1, p2 types.PoolI) sdk.Dec {
```

File: x/gamm/keeper/invariants.go [hunk 3]
```diff
@@ -78,3 +86,65 @@ func PoolTotalWeightInvariant(keeper Keeper, bk types.BankKeeper) sdk.Invariant
...
+	}
+
+	return product
+}
+
+var (
+	errorMargin, _ = sdk.NewDecFromStr("0.0001") // 0.01%
+)
+
+func PoolProductConstantInvariant(keeper Keeper) sdk.Invariant {
```
Comment (ValarDragon) [inline x/gamm/keeper/invariants.go:121]: Nice, this is a good way of doing it!

This should fail with liquidity removal though, correct?

File: x/gamm/keeper/math.go [hunk 4]
```diff
@@ -199,96 +193,12 @@ func calcPoolInGivenSingleOut(
...
-	} else {
-		return b.Sub(a), true
-	}
-}
-
-// pow computes base^(exp)
-// However since the exponent is not an integer, we must do an approximation algorithm.
-// TODO: In the future, lets add some optimized routines for common exponents, e.g. for common wIn / wOut ratios
-// Many simple exponents like 2:1 pools
-func pow(base sdk.Dec, exp sdk.Dec) sdk.Dec {
```
Comment (ValarDragon) [inline x/gamm/keeper/math.go:215]: Can we avoid moving this for now, and just make the function public?

File: x/gamm/simulation/operations.go [hunk 5]
```diff
@@ -144,14 +145,20 @@ func SimulateMsgCreatePool(ak stakingTypes.AccountKeeper, bk stakingTypes.BankKe
 		poolParams := genPoolParams(r, ctx.BlockTime(), poolAssets)
 
 		// TODO: Replace []string{} with all token types on chain.
-		futurePoolGovernor := genFuturePoolGovernor(r, simAccount.Address, []string{})
```
Comment (ValarDragon) [inline x/gamm/simulation/operations.go:147]: Why was this removed?

File: x/gamm/simulation/operations.go [hunk 6]
```diff
@@ -173,28 +180,30 @@ func SimulateMsgSwapExactAmountIn(ak stakingTypes.AccountKeeper, bk stakingTypes
 		}
 
 		coin := simCoins[r.Intn(len(simCoins))]
-		amt, _ := simtypes.RandPositiveInt(r, coin.Amount)
+		amt, _ := simtypes.RandPositiveInt(r, coin.Amount.QuoRaw(200))
```
Comment (ValarDragon) [inline x/gamm/simulation/operations.go:183]: Lets add comments for this

File: x/gamm/keeper/invariants.go (hunk 3 above)
Comment (ValarDragon) [reply x/gamm/keeper/invariants.go:121]: Also can you comment the function

File: x/gamm/simulation/operations.go [hunk 7]
```diff
@@ -204,17 +213,25 @@ func SimulateMsgSwapExactAmountIn(ak stakingTypes.AccountKeeper, bk stakingTypes
...
 	if len(pools) == 0 {
 		return
 	}
 
-	res = make([]types.SwapAmountInRoute, routeLen)
-	for i := range res {
-		for {
+	res = []types.SwapAmountInRoute{}
+	for i := 0; i < routeLen; i++ {
+		for retry := 0; retry < 10; retry++ {
```
Comment (ValarDragon) [inline x/gamm/simulation/operations.go:234]: Whats this retry for?

File: x/gamm/simulation/operations.go (hunk 5 above)
Comment (mconcat) [reply x/gamm/simulation/operations.go:147]: There was a problem with having an empty denom array as an argument on `r.Intn(len(denoms))`. Previously the gamm operations were not executed so this was not found.

File: x/gamm/keeper/invariants.go (hunk 3 above)
Comment (mconcat) [reply x/gamm/keeper/invariants.go:121]: Yes, it should fail with liquidity removal.

File: x/gamm/simulation/operations.go (hunk 7 above)
Comment (mconcat) [reply x/gamm/simulation/operations.go:234]: We need to find a pool that contains the input coin as an asset, but the randomly selected pool may not.

File: x/gamm/keeper/invariants.go (hunk 3 above)
Comment (ValarDragon) [reply x/gamm/keeper/invariants.go:121]: Lets add a coment explicitly stating that, and a TODO for ourselves to revisit how to do this with liquidity removal? (Perhaps we have a simulator-only variable for trackign amount of liquidity removed since last invokation)

**Comment by ValarDragon** [comment]
LGTM pending moving math functions back, and adding comments

**Comment by mconcat** [comment]
All applied!

File: x/gamm/simulation/operations.go (hunk 6 above)
Comment (ValarDragon) [reply x/gamm/simulation/operations.go:187]:
```suggestion
             // TODO: Make like a 33% probability of using a ton of balance 
		amt, _ := simtypes.RandPositiveInt(r, coin.Amount.QuoRaw(200))
```

**Review by ValarDragon — APPROVED** [review]
